mne/preprocessing/annotate_amplitude.py

```diff
@@ -0,0 +1,200 @@
+# License: BSD-3-Clause
+
+import numpy as np
+
+from ..io import BaseRaw
+from ..annotations import Annotations, _adjust_onset_meas_date
+from ..io.pick import _picks_to_idx, _picks_by_type, _get_channel_types
+from ..utils import _validate_type, verbose, logger, _mask_to_onsets_offsets
+
+
+@verbose
+def annotate_amplitude(raw, peak=None, flat=None, bad_percent=5,
+                       min_duration=0.005, picks=None, *, verbose):
+    """Annotate segments of raw data which PTP amplitudes exceeds thresholds
+    in ``peak`` or fall below thresholds in ``flat``.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data.
+    peak : float | dict | None
+        Reject segments based on **maximum** peak-to-peak signal amplitude
+        (PTP). Valid **keys** can be any channel type present in the object.
+        The **values** are floats that set the minimum acceptable PTP. If the
+        PTP is smaller than this threshold, the segment will be annotated.
+        If float, the minimum acceptable PTP is applied to all channels.
+    flat : float | dict | None
+        Reject segments based on **minimum** peak-to-peak signal amplitude
+        (PTP). Valid **keys** can be any channel type present in the object.
+        The **values** are floats that set the minimum acceptable PTP. If the
+        PTP is smaller than this threshold, the segment will be annotated.
+        If float, the minimum acceptable PTP is applied to all channels.
+    bad_percent : float
+        The percentage of the time a channel can be bad.
+        Below this percentage, temporal bad marking (:class:`~mne.Annotations`)
+        will be used. Above this percentage, spatial bad marking
+        (:class:`info['bads'] <mne.Info>`) will be used.
+        Defaults to 5 (5%%).
+    min_duration : float
+        The minimum duration (sec) to consider as above or below threshold.
+        For some systems with low bit data representations, adjacent
+        time samples with exactly the same value are not totally uncommon.
+        Defaults to 0.005 (5 ms).
+    %(picks_good_data)s
+    %(verbose)s
+
+    Returns
+    -------
+    annot : instance of Annotations
+        The annotated bad segments.
+    bads : list
+        The channels detected as bad.
+
+    Notes
+    -----
+    This function may perform much faster if data are loaded in memory, as it
+    loads data one channel at a time (across all time points), which is
+    typically not an efficient way to read raw data from disk.
```

Comment:
please check this. Cf comment below

Reply:
So, this is copy/pasted from the annotation flat method. It is probably still True to some extend. In `flat.py`, the function checks channel per channel; and loads data channel per channel. In this one, it loads data per channel type and performs most checks per channel type, except the rejection of short segments which is still done channel per channel (but doesn't involve data I/O, so irrelevant to this note).

I will test this and adapt the note accordingly later.

Reply:
I slightly changed it; it's still True to some extent, but it's not a performance bottleneck as it was with `annotate_flat`. It could still have some impact on large files with many different channel types. Let me know if you prefer to get rid of this note.